### DISCUSSION TITLE
How to run a gem5 arm aarch64 full system simulation with ruby_fs.py with more than 8 cores?

### DISCUSSION THREAD

**Discussion by duqi-hnu** [gem5/gem5#821] [Q&A] [answered]
In version v23, how to run a gem5 arm aarch64 full system simulation with ruby_fs.py with more than 8 cores?

My command argument is,  "./build/ARM/gem5.opt configs/example/arm/ruby_fs.py --kernel=fs/binaries/vmlinux.arm64 --disk-image=fs/ubuntu-18.04-arm64-docker.img --cpu o3 -n 16 --mem-size=32GB"

The error information is as follows:
"src/dev/arm/gic_v2.hh:375: fatal: fatal condition ctx >= 8 occurred: system.realview.gic requires the gem5_extensions parameter to support more than 8 cores"

**Comment by giactra** [comment, marked as answer]
Hi, this is due to the inherent limitations of GICv2 which supports 8 CPUs only. You have two solutions in front of you:

1. Use the gem5-GIC extension which allows to use GICv2 with up to 255 CPUs. To enable it you need to:
    - Set the GicV2 gem5_extensions param to True
    - Make sure the Linux GIC driver contains the gem5 specific patch removing the 8CPUs limitations [1]

2) Use a GICv3 powered platform (**recommended solution**). For example you could use the VExpress_GEM5_Foundation. At the moment you cannot change the machine type from command line for ruby_fs.py. I will open a PR to add this capability. You can however set the platform explicitly by tweaking the python config and pass the platform to the ArmRubySystem constructor [2] similarly to what we already do for fs_bigLITTLE.py [3]

[1]: https://gem5.googlesource.com/arm/linux/+/a2e8bda41866718ade4856d31f6c87e7234a3add%5E%21/
[2]: https://github.com/gem5/gem5/blob/stable/configs/example/arm/ruby_fs.py#L113
[3]: https://github.com/gem5/gem5/blob/stable/configs/example/arm/fs_bigLITTLE.py#L149C9-L149C17

**Reply by duqi-hnu** [reply]
Thank you.
I use VExpress_GEM5_V2 according to your method, and there is no error so far. However, 16 hours have passed, and the operating system has not been entered, and the memory usage has not increased through the htop command, and has remained at 1850M. Is this normal?
The following information is displayed by the htop command:
![1706755425807](https://github.com/gem5/gem5/assets/149283470/ce616199-6429-4223-a7a6-831e0a921d7f)

**Reply by duqi-hnu** [reply]
lose
Finally, an error was reported, and the error message was as follows:
src/dev/arm/gic_v3_distributor.cc:913: warn: Gicv3Distributor::write(): setting ARE_NS to 0 is not supported!
src/sim/port.cc:62: fatal: system.realview.gic.its.dma: Unconnected port!

**Reply by duqi-hnu** [reply]
This can be achieved using the first way, namely with the gem5-GIC extension, which allows GICv2 to use up to 255 cpus. @giactra，thank you very much

**Comment by giactra** [comment]
Hi, even if this questions has been marked as solved, I would still like to give an explanation to the error you are getting in case someone else wanted to go through the GICv3 route. 

> lose Finally, an error was reported, and the error message was as follows: src/dev/arm/gic_v3_distributor.cc:913: warn: Gicv3Distributor::write(): setting ARE_NS to 0 is not supported!

This is a "common" warning to be disregarded. It is generally speaking not an issue

> src/sim/port.cc:62: fatal: system.realview.gic.its.dma: Unconnected port!

This is where simulation fails (without even starting). The issue is that VExpress_GEM5_V2 is using the ITS (Interrupt Translation Service) of GICv3. This contains a dma port and what is happening is that you are not binding the port to any bus/interconnect/cache (meaning the ITS is left unconnected). You probably just need to hook it up.
Alternatively, the VExpress_GEM5_Foundation is not using the ITS
